### DISCUSSION TITLE
Example on how to store more complex data than string in state

### DISCUSSION THREAD

**Discussion by tymfear** [tauri-apps/tauri#7279] [Q&A] [answered]
Hi!

I've got a json with some content like
```json
{
  "token": "some string",
  "user": { 
      // lost of different fields here
  }
}
```

I would like to: 
- store data in state 
- be able to read and modify data.
- be able to check some of the fields in that `user` part, which might be missing (and of course I wouldn't want to describe tons of fields in structs)
- have state that be either empty or with that json/object.
In TS it would be something like
```typescript
interface State {
 token: string;
 user: Record<string, any>
}

let state: State | undefined;
```

Documentation won't help here as it has an example with a string.
As a Rust newcomer - I have no idea of what I should do.
There suppose to be some traits, structs and stuff that I have no idea of how to do.
Watched a lot of videos and read articles, but that's not enough, coz still lack of hands-on experience with Rust.

Could someone give a brief example of how to achieve that?

**Comment by FabianLars** [comment, marked as answer]
first of all, there's a slightly more extensive example here: <https://docs.rs/tauri/latest/tauri/struct.Builder.html#method.manage> which doesn't use just a string. Anyway, let's break down the points and then put it together at the end:
- store data in state
  - Alright, that's the base so let's ignore that for now
- be able to read and modify data
  - reading is supported as-is. modifying data needs interior mutability, in this case i'll use a [Mutex](https://doc.rust-lang.org/std/sync/struct.Mutex.html)
- be able to check some of the fields in that user part, which might be missing (and of course I wouldn't want to describe tons of fields in structs)
  - If you don't want to have all possible struct members predefined we'll need to use [`serde_json::Value`](https://docs.rs/serde_json/latest/serde_json/enum.Value.html), either by itself, or inside a hashmap (i'll do the latter since it makes accessing the outer most layer easier)
- have state that be either empty or with that json/object.
  - If we're talking about the whole state object this will involve an `Option`. If it'd be just about some parts of the state serde_json::Value would handle that too.

Alright, let's put it together. Note that i will use `Result`s in the commands even though they are only really needed for async commands to make it easier to convert them if needed. I also use a `type` alias for the state to make the code easier to read.

```rs
use serde::{Deserialize, Serialize};
use serde_json::Value;
use tauri::State;

use std::{collections::HashMap, sync::Mutex};

type WrappedState = Mutex<Option<MyState>>;

#[derive(Clone, Debug, Serialize, Deserialize)]
struct MyState {
    token: String,
    user: HashMap<String, Value>,
}

/// Actually init the state with some value
#[tauri::command]
fn init_state(state: State<WrappedState>, new_state: MyState) -> Result<(), ()> {
    // This will replace the whole state object so you may have to add a check if you don't want that.
    *state.lock().unwrap() = Some(new_state);

    Ok(())
}

/// Get the whole state object to the frontend
#[tauri::command]
fn get_state(state: State<WrappedState>) -> Result<Option<MyState>, ()> {
    // This will return a clone (deep copy?) of the stored value if any, otherwise None/null
    Ok(state.lock().unwrap().clone())
}

/// Get the a single entry from the user hashmap
#[tauri::command]
fn get_user_entry(state: State<WrappedState>, key: String) -> Result<Option<Value>, ()> {
    // We're explicitly using `ref` to prevent Rust from trying to `Copy` the Option.
    if let Some(ref state) = *state.lock().unwrap() {
        // Note we use `cloned` instead of `clone` here to convert `Option<&Value>` to a cloned `Option<Value>`.
        // clone would only return a cloned `Option<&Value>`.
        return Ok(state.user.get(&key).cloned());
    }

    // return None if there's no state. You could also return an error if you prefer.
    Ok(None)
}

/// Modify a single entry from the user hashmap
#[tauri::command]
fn set_user_entry(state: State<WrappedState>, key: String, val: Value) -> Result<(), ()> {
    // Note the `mut` keyword for the first time here.
    // You'd need that in `init_state` too if you would bind the MutexGuard to a variable.
    if let Some(ref mut state) = *state.lock().unwrap() {
        state.user.insert(key, val);
    }

    // Always returning Ok() is a weird choice here.
    // You probably want to either return an error asking the frontend to first set a state if there was none,
    // or init the state with some base values in rust if that's possible.
    Ok(())
}

fn main() {
    // We init the state here to make it easier to work with later.
    // This is optional but you'll have to use the AppHandle's `try_state` to receive the state in your commands and can't just inject them via `tauri::State`
  tauri::Builder::default()
    .manage(Mutex::new(None::<MyState>))
    .invoke_handler(tauri::generate_handler![init_state, get_state, get_user_entry, set_user_entry])
    .run(tauri::generate_context!())
    .expect("error while running tauri app");
}
```

I hope this makes it a bit clearer and please feel free to hit me with criticism. I could use feedback on this to better understand what we should put into the explicit State guide (that does not exist yet)

**Reply by tymfear** [reply]
Thank you very much for an example! 
As a feedback - it won't compile for me, getting the following
```
error[E0282]: type annotations needed
  --> src/main.rs:66:28
   |
66 |         .manage(Mutex::new(None))
   |                            ^^^^ cannot infer type of the type parameter `T` declared on the enum `Option`
   |
help: consider specifying the generic argument
   |
66 |         .manage(Mutex::new(None::<T>))
   |                                +++++
```
Would you be able to advise some fix for that.
And once it compiled - you could just add it to docs as is - sure that it would be really helpful!

**Reply by FabianLars** [reply]
Of course I messed up the only line i did not write in vscode but in GitHub directly...
I'm only on my phone right now so I can't test it but can you try this?
```
.manage(Mutex::new(None::<MyState>))
```

**Reply by tymfear** [reply]
Now it's 
```
error[E0507]: cannot move out of dereference of `MutexGuard<'_, std::option::Option<MyState>>`
  --> src/main.rs:37:26
   |
37 |     if let Some(state) = *state.lock().unwrap() {
   |                 -----    ^^^^^^^^^^^^^^^^^^^^^^
   |                 |
   |                 data moved here
   |                 move occurs because `state` has type `MyState`, which does not implement the `Copy` trait
   |
help: consider removing the dereference here
   |
37 -     if let Some(state) = *state.lock().unwrap() {
37 +     if let Some(state) = state.lock().unwrap() {
   |

```

and same for later
```
error[E0507]: cannot move out of dereference of `MutexGuard<'_, std::option::Option<MyState>>`
  --> src/main.rs:52:30
   |
52 |     if let Some(mut state) = *state.lock().unwrap() {
   |                 ---------    ^^^^^^^^^^^^^^^^^^^^^^
   |                 |
   |                 data moved here
   |                 move occurs because `state` has type `MyState`, which does not implement the `Copy` trait
   |
help: consider removing the dereference here
   |
52 -     if let Some(mut state) = *state.lock().unwrap() {
52 +     if let Some(mut state) = state.lock().unwrap() {
   |
```

Tried to add derive Copy for `MyState` - didn't work either

**Reply by FabianLars** [reply]
Okay now that one is weird cause it didn't complain about that on my end when I wrote that code. I'll check again on my pc tomorrow morning - probably should have waited till tomorrow to post this in the first place to not make these sleep deprived mistakes 😂

**Reply by FabianLars** [reply]
Did you try just removing the `*` there like the error says?

**Reply by tymfear** [reply]
```
  --> src/main.rs:37:12
   |
37 |     if let Some(state) = state.lock().unwrap() {
   |            ^^^^^^^^^^^   --------------------- this expression has type `MutexGuard<'_, std::option::Option<MyState>>`
   |            |
   |            expected `MutexGuard<'_, Option<MyState>>`, found `Option<_>`
   |
   = note: expected struct `MutexGuard<'_, std::option::Option<MyState>, >`
                found enum `std::option::Option<_>`
```

**Reply by FabianLars** [reply]
I updated the script in my initial comment with something that actually compiles for me (it failed silently on an unrelated issue before so i didn't see all code errors :/ ). Sorry for the back and forth here.

**Reply by tymfear** [reply]
New example works - there's small warning about unnecessary `mut` in `set_user_entry` but otherwise example compiles and works. Thank you very much. Imo this example should be somewhere in docs.

Also - do you happen to have example somewhere on how to extract commands into separate file? Like put commands into commands.rs or something?

**Reply by FabianLars** [reply]
https://github.com/tauri-apps/tauri/tree/dev/examples/commands

Edit: Removed the `mut` btw

**Reply by oleghalin** [reply]
Trying you use your sample above but for some reason, tauri throws an error in the command that I should run .manage before binding state. I replace `Mutex::new(None::<ApplicationState>)` to `Mutex::new(ApplicationState::default())`. Maybe you can point me to what am I doing wrong? 

UPD: i found out that my WrapperState is referencing to Option<>, so as soon as i wrapped my ::default to Option everything goes well
